File: LeetCode_ProblemSet/Problem368.cpp
#include <algorithm>
#include <functional>
#include <vector>

class Solution {
   public:
    std::vector<int> largestDivisibleSubset(std::vector<int> &nums) {
        int n = nums.size();

        std::vector<std::pair<int, int>> numsWithIndices(n);
        for (int i = 0; i < n; i++) {
            numsWithIndices[i] = {nums[i], i};
        }

        std::sort(numsWithIndices.begin(), numsWithIndices.end());

        std::vector<int> opt(n, 0);
        std::vector<int> parent(n, -1);
        opt[numsWithIndices[0].second] = 1;
        for (int i = 1; i < n; i++) {
            const auto &[num1, idx1] = numsWithIndices[i];

            for (int j = 0; j < i; j++) {
                const auto &[num2, idx2] = numsWithIndices[j];

                if (num1 % num2 != 0) continue;

                if (1 + opt[idx2] > opt[idx1]) {
                    opt[idx1] = 1 + opt[idx2];
                    parent[idx1] = idx2;
                }
            }

            if (opt[idx1] == 0) opt[idx1] = 1;
        }

        int maxIdx = -1, maxSize = 0;
        for (int i = 0; i < n; i++) {
            if (opt[i] > maxSize) {
                maxSize = opt[i];
                maxIdx = i;
            }
        }

        std::vector<int> answer;
        answer.reserve(n);

        int curIdx = maxIdx;
        while (curIdx != -1) {
            answer.push_back(nums[curIdx]);
            curIdx = parent[curIdx];
        }

        return answer;
    }
};

File: LeetCode_ProblemSet/Rust/src/Problem368.rs
impl Solution {
    pub fn largest_divisible_subset(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();

        let mut num_with_indices: Vec<(i32, usize)> = nums
            .iter()
            .enumerate()
            .map(|(idx, &num)| (num, idx))
            .collect();
        num_with_indices.sort();

        let mut opt = vec![0; n];
        let mut parent = vec![n; n];
        opt[num_with_indices[0].1] = 1;

        for i in 1..n {
            let (num1, idx1) = num_with_indices[i];

            for j in 0..i {
                let (num2, idx2) = num_with_indices[j];
                if num1 % num2 != 0 {
                    continue;
                }
                if opt[idx2] + 1 > opt[idx1] {
                    opt[idx1] = 1 + opt[idx2];
                    parent[idx1] = idx2;
                }
            }

            if opt[idx1] == 0 {
                opt[idx1] = 1;
            }
        }

        let (mut max_idx, mut max_size) = (n, 0);
        for i in 0..n {
            if opt[i] > max_size {
                max_size = opt[i];
                max_idx = i;
            }
        }

        let mut answer = Vec::new();
        let mut cur_idx = max_idx;
        while cur_idx != n {
            answer.push(nums[cur_idx]);
            cur_idx = parent[cur_idx];
        }

        return answer;
    }
}
